modules/imgproc/src/intelligent_scissors.cpp

```diff
@@ -0,0 +1,159 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2018-2020, Intel Corporation, all rights reserved.
+// Third party copyrights are property of their respective owners.
+
+
+#include "precomp.hpp"
+
+#include <queue>
+
+namespace cv
+{
+struct Pix
+{
+    Point next_point;
+    double cost;
+
+    bool operator > (const Pix &b) const
+    {
+        return cost > b.cost;
+    }
+};
+
+static float lcost(const Point& p, const Point& q, const Mat& gradient_magnitude, const Mat& Iy, const Mat& Ix, const Mat& zero_crossing)
+{
+    float fG = gradient_magnitude.at<float>(q.y, q.x);
+    float dp;
+    float dq;
+    const float WEIGHT_LAP_ZERO_CROSS = 0.43f;
+    const float WEIGHT_GRADIENT_MAGNITUDE = 0.14f;
+    const float WEIGHT_GRADIENT_DIRECTION = 0.43f;
+    bool isDiag = (p.x != q.x) && (p.y != q.y);
+
+    if ((Iy.at<float>(p) * (q.x - p.x) - Ix.at<float>(p) * (q.y - p.y)) >= 0)
+    {
+        dp = Iy.at<float>(p) * (q.x - p.x) - Ix.at<float>(p) * (q.y - p.y);
+        dq = Iy.at<float>(q) * (q.x - p.x) - Ix.at<float>(q) * (q.y - p.y);
+    }
+    else
+    {
+        dp = Iy.at<float>(p) * (p.x - q.x) + (-Ix.at<float>(p)) * (p.y - q.y);
+        dq = Iy.at<float>(q) * (p.x - q.x) + (-Ix.at<float>(q)) * (p.y - q.y);
+    }
+    if (isDiag)
+    {
+        dp /= sqrtf(2);
+        dq /= sqrtf(2);
+    }
+    else
+    {
+        fG /= sqrtf(2);
+    }
+    return  WEIGHT_LAP_ZERO_CROSS * zero_crossing.at<uchar>(q) +
+            WEIGHT_GRADIENT_DIRECTION * (acosf(dp) + acosf(dq)) / static_cast<float>(CV_PI) +
+            WEIGHT_GRADIENT_MAGNITUDE * fG;
+}
+
+class IntelligentScissorsImpl : public IntelligentScissors
+{
+public:
+    void apply(InputArray img, OutputArray total_hit_map_x, OutputArray total_hit_map_y, const Point start_point) CV_OVERRIDE
+    {
+        const int EDGE_THRESHOLD_LOW = 50;
+        const int EDGE_THRESHOLD_HIGH = 100;
+        Mat src = img.getMat();
+
+        total_hit_map_x.create(src.size(), CV_32SC1);
+        Mat hit_map_x = total_hit_map_x.getMat();
+
+        total_hit_map_y.create(src.size(), CV_32SC1);
+        Mat hit_map_y = total_hit_map_y.getMat();
+
+        Mat grayscale, img_canny, Ix, Iy;
+        Mat zero_crossing, gradient_magnitude;
+
+        cvtColor(src, grayscale, COLOR_BGR2GRAY);
+        Canny(grayscale, img_canny, EDGE_THRESHOLD_LOW, EDGE_THRESHOLD_HIGH);
+        threshold(img_canny, zero_crossing, 254, 1, THRESH_BINARY_INV);
```

Comment:
`zero_crossing` computation does not follow the article. Canny() is not used there.

See Section 3.1